### PR TITLE
Backup/Restore

### PR DESCRIPTION
**Pull request by cscetbon**
| Q               | A
| --------------- | ---
| New feature?    | no|yes

### What's in this PR?

Add Backup/Restore ability

### CODE REVIEW

File: Readme.md [hunk 1]
```diff
@@ -35,7 +34,79 @@ For detailed installation instructions, see the [Casskop Documentation Page](htt
 
 The documentation of the Casskop operator project is available at the [Casskop Documentation Page](https://orange-opensource.github.io/casskop/docs/1_concepts/1_introduction).
 
-## Cassandra operator
+
+CassKop deals with Cassandra clusters on one datacenter. For multi-dacenters deployments, please use [MultiCassKop](multi-casskop/Readme.md) in addition to CassKop. This second operator is part of this same repository.
```
Comment (erdrix) [inline Readme.md:38]: This part wasn't removed ?

File: pkg/controller/cassandrarestore/reconcile.go [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+	typeMeta := restore.TypeMeta
+	err := r.client.Update(ctx, restore)
+	if err != nil {
+		return nil, err
+	}
+	restore.TypeMeta = typeMeta
+	return restore, nil
+}
+
+func (r *ReconcileCassandraRestore) listPods(namespace string, selector map[string]string) (*v1.PodList, error) {
```
Comment (erdrix) [inline pkg/controller/cassandrarestore/reconcile.go:295]: this code is never re-used somewhere else ?

File: pkg/controller/common/common_controller.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	return reconcile.Result{}, err
+}
+
+// reconciled returns an empty result with nil error to signal a successful reconcile
+// to the controller manager
+func Reconciled() (reconcile.Result, error) {
+	return reconcile.Result{}, nil
+}
+
+// newSidecarsConnection is a convenience wrapper for creating a sidecars connection
```
Comment (erdrix) [inline pkg/controller/common/common_controller.go:29]: function renamed !

File: pkg/controller/cassandrarestore/reconcile.go (hunk 2 above)
Comment (cscetbon) [reply pkg/controller/cassandrarestore/reconcile.go:295]: there are identical functions but on different reconcile object. Maybe we should refactor it under as they all use the client of the reconcile object ?

File: Readme.md (hunk 1 above)
Comment (cscetbon) [reply Readme.md:38]: Correct, I don't know why it's removed. Adding it back

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
+| Q               | A
+| --------------- | ---
+| Bug fix?        | no|yes
+| New feature?    | no|yes
+| API breaks?     | no|yes
+| Deprecations?   | no|yes
```
Comment (cscetbon) [inline .github/PULL_REQUEST_TEMPLATE.md:6]: I think we can replace them by checkboxes instead

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
+# How to contribute
+
+CassKop is Apache 2.0 licensed and accepts contributions via GitHub pull requests. This document outlines some of the
+conventions on commit message formatting, contact points for developers, and other resources to help get contributions
+into CassKop.
+
+## Email and Chat
+
```
Comment (cscetbon) [inline CONTRIBUTING.md:8]: There is no email. Should we keep only Slack ? I think it makes sense

File: Readme.md [hunk 6]
```diff
@@ -1,14 +1,16 @@
+<img src="static/casskop.png" alt="Logo" width="150"/>
+
 # CassKop - Cassandra Kubernetes operator
 
-[![CircleCI](https://circleci.com/gh/Orange-OpenSource/cassandra-k8s-operator.svg?style=svg&circle-token=480ca5c31a9e9ef9b893151dd2d7c15eaf0e94d0)](https://circleci.com/gh/Orange-OpenSource/cassandra-k8s-operator)
+[![CircleCI](https://circleci.com/gh/Orange-OpenSource/casskop.svg?style=svg&circle-token=480ca5c31a9e9ef9b893151dd2d7c15eaf0e94d0)](https://circleci.com/gh/Orange-OpenSource/casskop) [![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=Orange-OpenSource_cassandra-k8s-operator&metric=alert_status)](https://sonarcloud.io/dashboard?id=Orange-OpenSource_cassandra-k8s-operator)
```
Comment (cscetbon) [inline Readme.md:5]: It's not a private repository. We shouldn't need a token

File: cassandra-stress/cassandra-stress-medium.yaml [hunk 7]
```diff
@@ -14,18 +14,18 @@ spec:
...
   - name: cassie1-cassandra-stress
-    image: orangeopensource/cassandra-image
+    image: cassandra
     imagePullPolicy: "IfNotPresent"
     securityContext:
       capabilities:
         add: ["IPC_LOCK"]
     command: ["/bin/sh"]
-    args: ["-c", "cassandra-stress 'user profile=/opt/cassandra-stress/medium_stress.yaml ops(insert=1) n=5m cl=one -node cassandra-demo-dc1.cassandra-demo -mode native cql3 user=bench password=monbench -rate threads=50 -pop seq=0..5m -graph file=/tmp/stress-medium1.html' && echo END && while true ; do sleep 60; done"]
+    args: ["-c", "cassandra-stress 'user profile=/opt/cassandra-stress/medium_stress.yaml ops(insert=1,query_by_sub_id=1) n=5m cl=one -node cassandra-demo -mode native cql3 user=cassandra password=cassandra -rate threads=50 -pop seq=0..5m -graph file=/tmp/stress-medium1.html' && echo END && while true ; do sleep 60; done"]
```
Comment (cscetbon) [inline cassandra-stress/cassandra-stress-medium.yaml:28]: why add `query_by_sub_id`? what about the other files ?

File: CONTRIBUTING.md (hunk 5 above)
Comment (erdrix) [reply CONTRIBUTING.md:8]: You can add the e-mail : prj.casskop.support@list.orangeportails.net

File: pkg/controller/cassandrabackup/backup.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+		corev1.EventTypeNormal,
+		"BackupInitiated",
+		fmt.Sprintf("Task initiated to backup datacenter %s of cluster %s to %s under snapshot %s",
+			backupClient.backup.Spec.Datacenter, backupClient.backup.Spec.CassandraCluster,
+			backupClient.backup.Spec.StorageLocation, backupClient.backup.Spec.SnapshotTag))
+
+	for range time.NewTicker(2 * time.Second).C {
+		if status, err := backrestClient.BackupStatusByID(operationID); err != nil {
+			logging.Error(err, fmt.Sprintf("Error while finding submitted backup operation %v", operationID))
+			break
```
Comment (erdrix) [inline pkg/controller/cassandrabackup/backup.go:52]: Do we really want to break the loop ? 
I mean if the error is due to the operation not found : ok,  but if it's just a timeout or something else, the next call might work, so why not try again ?

File: pkg/controller/cassandrabackup/backup.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+		} else {
+			backupClient.updateStatus(status, logging)
```
Comment (erdrix) [inline pkg/controller/cassandrabackup/backup.go:54]: You might be in a case that the Backup update failed, and you only log the error, but the object will not be updated :/
Shouldn't we catch the error and try again on the next iteration ?

File: pkg/controller/cassandrabackup/reconcile.go [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+
+		// Add Finalizer
+		r.setFinalizer(cassandraBackup, true)
+	}
+
+	if cassandraBackup.IsScheduled() {
+		return reconcile.Result{}, nil
+	}
+
+	return reconcile.Result{}, r.backupData(cassandraBackup, cc, reqLogger)
```
Comment (erdrix) [inline pkg/controller/cassandrabackup/reconcile.go:228]: What happens if we apply changes to a CassandraBackup and a backup is already running ? 
Is it starting another backup ? Does it kill the one who is running ? Does it do nothing ?

File: pkg/controller/cassandrabackup/backup.go (hunk 8 above)
Comment (cscetbon) [reply pkg/controller/cassandrabackup/backup.go:52]: @erdrix I don't think it makes sense. Only 2 possible errors https://github.com/Orange-OpenSource/casskop/blob/backup-operator/pkg/cassandrabackup/operations.go#L55-L74 and both of them are not recoverable

File: pkg/controller/cassandrabackup/reconcile.go (hunk 10 above)
Comment (cscetbon) [reply pkg/controller/cassandrabackup/reconcile.go:228]: @erdrix If the backup is scheduled, it's updated. If it's not, no new backup runs https://github.com/Orange-OpenSource/casskop/blob/backup-operator/pkg/controller/cassandrabackup/reconcile.go#L120-L121

**Review by erdrix — DISMISSED** [review]

**Pull request merged**